### PR TITLE
Jpeg 9 compatibility

### PR DESCRIPTION
**Pull request by 183amir**

### CODE REVIEW

**Comment by 183amir** [comment]
@pelson I am getting a very wired error on this package. Could you look into it please?
ldd'ing the _library.so gives me this:

```
ldd /home/amir/miniconda/envs/_test/lib/python3.5/site-packages/bob/io/image/_library.cpython-35m-x86_64-linux-gnu.so 
    linux-vdso.so.1 (0x00007ffdaabc4000)
    libbob_core.so => not found
    libbob_io_base.so => not found
    libboost_system.so.1.60.0 => /home/amir/miniconda/envs/_test/lib/python3.5/site-packages/bob/io/image/../../../../../libboost_system.so.1.60.0 (0x00007f17c8aa7000)
    libboost_filesystem.so.1.60.0 => /home/amir/miniconda/envs/_test/lib/python3.5/site-packages/bob/io/image/../../../../../libboost_filesystem.so.1.60.0 (0x00007f17c888e000)
    libblitz.so.0 => /home/amir/miniconda/envs/_test/lib/python3.5/site-packages/bob/io/image/../../../../../libblitz.so.0 (0x00007f17c868a000)
    libpng16.so.16 => /home/amir/miniconda/envs/_test/lib/python3.5/site-packages/bob/io/image/../../../../../libpng16.so.16 (0x00007f17c8448000)
***HERE*** libjpeg.so.9 => /home/amir/miniconda/envs/_test/lib/python3.5/site-packages/bob/io/image/../../../../../libjpeg.so.9 (0x00007f17c820c000)
    libtiff.so.5 => /home/amir/miniconda/envs/_test/lib/python3.5/site-packages/bob/io/image/../../../../../libtiff.so.5 (0x00007f17c7f9d000)
    libgif.so.7 => /home/amir/miniconda/envs/_test/lib/python3.5/site-packages/bob/io/image/../../../../../libgif.so.7 (0x00007f17c7d94000)
    libpython3.5m.so.1.0 => /home/amir/miniconda/envs/_test/lib/python3.5/site-packages/bob/io/image/../../../../../libpython3.5m.so.1.0 (0x00007f17c78a5000)
    libstdc++.so.6 => /home/amir/miniconda/envs/_test/lib/python3.5/site-packages/bob/io/image/../../../../../libstdc++.so.6 (0x00007f17c758e000)
    libm.so.6 => /usr/lib/libm.so.6 (0x00007f17c7289000)
    libgcc_s.so.1 => /home/amir/miniconda/envs/_test/lib/python3.5/site-packages/bob/io/image/../../../../../libgcc_s.so.1 (0x00007f17c7073000)
    libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007f17c6e55000)
    libc.so.6 => /usr/lib/libc.so.6 (0x00007f17c6ab4000)
    librt.so.1 => /usr/lib/librt.so.1 (0x00007f17c68ac000)
    libz.so.1 => /home/amir/miniconda/envs/_test/lib/python3.5/site-packages/bob/io/image/../../../../.././libz.so.1 (0x00007f17c6695000)
***HERE*** libjpeg.so.62 => not found
    libdl.so.2 => /usr/lib/libdl.so.2 (0x00007f17c6490000)
    libutil.so.1 => /usr/lib/libutil.so.1 (0x00007f17c628d000)
    /usr/lib64/ld-linux-x86-64.so.2 (0x0000558089577000)
```

**Comment by 183amir** [comment]
@pelson why conda is not pulling `libtiff 4.0.6-3` here which is linked against `jpeg 9`?
If I pin the version to `libtiff 4.0.6`, it works.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -23,11 +25,12 @@ requirements:
...
   - giflib
-  - libpng
-  - jpeg
-  - libtiff
+  - libpng 1.6*
+  - jpeg 9*
+  - libtiff 4.0.*
   - pkg-config
-  - gcc     # [linux]
+  - isl 0.12.2  # [linux]
```
Comment (pelson) [inline recipe/meta.yaml]: This shouldn't be needed in the future. The gcc package should express this.

Comment (183amir) [inline recipe/meta.yaml]: Yes, I added it temporarily.

Comment (pelson) [inline recipe/meta.yaml]: We should probably add a comment along those lines (pointing back to https://github.com/conda-forge/staged-recipes/issues/436#issuecomment-212323816)

**Comment by jakirkham** [comment]
Go ahead and try restarting Circle CI.

**Comment by 183amir** [comment]
Back to original problem here after all that `gcc` problems:  why conda is not pulling libtiff 4.0.6-3 here which is linked against jpeg 9? If I pin the version to libtiff 4.0.6, it works.

**Comment by jakirkham** [comment]
@ocefpaf, I think you know more about what is going on here than I do. Sorry I haven't followed as closely as I probably should. Could you please help us understand how the pinning needs to be done here?

**Comment by ocefpaf** [comment]
> @ocefpaf, I think you know more about what is going on here than I do. Sorry I haven't followed as closely as I probably should. Could you please help us understand how the pinning needs to be done here?

Sorry, but things are really messy right now and the solution is not clear. In order to get some packages working on OS X I added `xz` (`libzma`) version `5.2.2`. All the packages that depend on `xz` got re-built and pinned to  `5.2.2`.

The problem is that Python 3.4 and 3.5 are built against `xz 5.0.5`. So conda downgrades the `xz` and ends up getting packages that are not built against `jpeg 9b` while installing `jpeg 9b`. (Maybe we have pinned something wrongly at some point.)

In `gdal` I am removing support to `openjpeg` (and its dependency `libtiff`) on `Python >3.4` while I wait for our Python 3.5 recipe to get merged.

The alternative is to pin everything to versions that use `xz 5.0.5` and `jpeg 8d`. (BTW that is not an option for `gdal` because then I cannot build OS X, but it can be an option here.)

**Comment by pelson** [comment]
> Maybe we have pinned something wrongly at some point.

Anything which doesn't pin jpeg is wrongly packaged at this point. We should remove those distributions from anaconda.org IMO.

**Comment by jakirkham** [comment]
> Anything which doesn't pin jpeg is wrongly packaged at this point.

This is tricky. Some of things are dependencies that might grind the build process to a halt.

Right now, I'm going to try to have `libtiff` shadow the `defaults` one (same version and build number). This is what we did with `cloog` and that seems to be working well. I think this will help us get out of the hole on that point. The `xz` stuff is another story I have not been able to dig into yet.

**Comment by ocefpaf** [comment]
> Right now, I'm going to try to have `libtiff` shadow the `defaults` one (same version and build number). This is what we did with `cloog` and that seems to be working well. I think this will help us get out of the hole on that point. The `xz` stuff is another story I have not been able to dig into yet.

The problem is that our `libtiff 4.0.6` (which is needed due to some  OS X crashes when using the default channel `libtiff 4.0.2`) is compiled against our `xz 5.2.2` instead of the default channel `xz 5.0.5`. The latter is used in Python 3.4 and 3.5 making everything that needs a modern `libtiff` or `xz` useless there.

**Comment by jakirkham** [comment]
Maybe you could try changing the pinning to `jpeg 8d`?

**Comment by pelson** [comment]
> Maybe you could try changing the pinning to jpeg 8d?

That would only be an intermediate step, of course. Once all of your dependencies have associated jpeg 9 builds, you would be able to move forwards.

**Comment by jakirkham** [comment]
Yeah, I worry that there is something subtle that we are missing here about this `jpeg` pinning.

**Comment by 183amir** [comment]
I already have a version built with `jpeg` 8 but its not pinned on it.

**Comment by ocefpaf** [comment]
Ping @jakirkham and @183amir. I re-started `CircleCI` and things are green now.

**Comment by 183amir** [comment]
Thank you @ocefpaf.

**Pull request merged**